From.pm: allow single-digit values.

jonbakke opened the pull request:
Karabiner's names for the digits are the digits themselves, so allow single-digit characters without triggering the 'Specifiers such as lazy, repeat, halt, and hold_down_in_milliseconds do not apply in "from" actions' croak (line 42).

jonbakke committed:
From.pm: allow single-digit values.

Karabiner's names for the digits are the digits themselves, so allow single-digit characters without triggering the 'Specifiers such as lazy, repeat, halt, and hold_down_in_milliseconds do not apply in "from" actions' croak (line 42).

sdondley reviewed (commented):
Thanks for you contribution. Can you please send a snippet of perl code that is not handled properly by the module to help jog my memory? Please also describe what was expected to happen and what actually happened.

jonbakke commented:
Of course! Thanks for looking this over.

### Example
```
use JSON::Karabiner::Manipulator;
set_title 'changing digits';
set_rule_name '1 is 2';
new_manipulator;
add_action 'from';
add_key_code '1';
add_action 'to';
add_key_code '2';
write_file;
```

### Expected behavior
Successful conversion by Karabiner.pm, the output of which is recognized by Karabiner-Elements, which allows adding the new complex rule '1 is 2', and thereafter presses of key 1 are recognized as presses of key 2.

This is the file generated with my proposed change, which appears to work as expected. (I haven't run any formal tests because that's outside my skill set. However, that's a skill I need to learn so if running/writing tests would help you, please let me know and I'll give it a try.)
```
{
   "rules" : [
      {
         "description" : "1 is 2",
         "manipulators" : [
            {
               "from" : {
                  "key_code" : "1"
               },
               "to" : [
                  {
                     "key_code" : "2"
                  }
               ],
               "type" : "basic"
            }
         ]
      }
   ],
   "title" : "changing digits"
}
```

### Actual behavior
Karabiner.pm 0.018 from cpanm, perl 5.32 from brew, and macOS 10.15.5.
(my escaped newlines added)
```
$ perl test.pl
Specifiers such as lazy, repeat, halt, and hold_down_in_milliseconds \
do not apply in "from" actions at /usr/local/Cellar/perl/5.32.0/lib/perl5/\
site_perl/5.32.0/JSON/Karabiner/Manipulator.pm line 116.
```

sdondley commented:
This is helpful. I vaguely remember that my way of indicating number of milliseconds was kind of a quick and dirty hack. I'll take a close look at this later this week. Thanks again.

jonbakke committed:
Update From.pm

I was unable to use JSON::Karabiner to produce rules for simultaneous keypresses. The below changes work for me, but I'm sure there are more elegant (and consistent) ways to achieve the same (or better) results.

### Example file
```
use JSON::Karabiner::Manipulator;
set_title 'boolean test';
set_rule_name 'a->b is c';
new_manipulator;
add_action 'from';
add_simultaneous 'a', 'b';
add_simultaneous_options 'detect_key_down_uninterruptedly', 'false';
add_action 'to';
add_key_code 'c';
write_file;
```

### Expected results
Compiled by JSON::Karabiner, loads in Karabiner-Elements, changes a key combo of 'a' and 'b' to 'c'.

This is an example of the desired output:
```
{
   "rules" : [
      {
         "description" : "a->b is c",
         "manipulators" : [
            {
               "from" : {
                  "simultaneous" : [
                     {
                        "key_code" : "a"
                     },
                     {
                        "key_code" : "b"
                     }
                  ],
                  "simultaneous_options" : {
                     "detect_key_down_uninterruptedly" : false,
                     "key_down_order" : "strict"
                  }
               },
               "to" : [
                  {
                     "key_code" : "c"
                  }
               ],
               "type" : "basic"
            }
         ]
      }
   ],
   "title" : "boolean test"
}
```

### Observed results
JSON::Karabiner fails to compile the example file with the following message:
```
Simultaneous option is not a valid option at /usr/local/Cellar/perl/5.32.0/lib/perl5/site_perl/5.32.0/JSON/Karabiner/Manipulator.pm line 116.
```

### Proposed fixes
In JSON/Karabiner/Manipulator/Actions/From.pm --
In sub add_simultaneous_options --

#### 1: initialization of $exists
Fix typo where what should be a test is an assignment: `$_ eq $option` not `$_ = $option`.

#### 2: standardize error message
The other $option croak messages specify what the offending $option value was, so it is included here as well.

This is unrelated to the main issue; I could split it into two PRs if that helps.

#### 3: testing $exists
Since the above assignment changed each element of @allowed_options to $option before assigning a value to $exists, the current test (`croak...if $exists`) will always be true. Instead, we want $option to match exactly one element of @allowed_options, so the test is changed to `croak...unless ($exists == 1)`.
#### 4: allow uninitialized $option
The `croak...if ($s->{"so${option}_is_set"} == 1)` will always test against an uninitialized value on the first iteration. The predominant idiom for similar tests in this software is to simply test its existence, so that pattern is maintained with `croak...if $s->{"so_${option}_is_set"}`.

#### 5: use JSON booleans
As is, the package will export true and false values as strings, which are rejected by Karabiner-Elements (`"error: ... must be boolean, but is "true"`). Therefore, while testing the validity of the proposed detect_key_down_uninterruptedly value, it is also assigned to `$JSON::true` or `$JSON::false`.

I suspect there is a reusable way to do this but I have not looked.

#### 6: implement settings
As is, the package only records whether these settings have been applied or not; they never actually change the JSON output. So `$s->{data}{simultaneous_options}` is given a value.

The construction of the if-defined block is ugly, but yields a single JSON object that both looks good in JSON and is accepted by Karabiner-Elements. Concatenating the hashes did not work for me (Karabiner-Elements: "error: json must be object, but is [{a:b},{c:d}]"). Another approach that appears to work is to output multiple simultaneous_options objects in the JSON file, but this does not appear convenient with JSON::Karabiner's internal data structure.

jonbakke committed:
Merge pull request #1 from jonbakke/add_simultaneous-1

From.pm: handle simultaneous keys
